### PR TITLE
8229186: Improve error messages for TestStringIntrinsics failures

### PR DESCRIPTION
**Pull request by lepestock**
pre-Skara RFR thread: [link](https://mail.openjdk.java.net/pipermail/hotspot-compiler-dev/2020-May/038416.html)

Error reporting was improved by writing a C-style escaped string representations for the variables passed to the methods being tested. For array comparisons, a dedicated diff-formatter was implemented.

Sample output for comparing byte arrays (with artificial failure):
``` ----------System.err:(21/1553)----------
 Result: (false) of 'arrayEqualsB' is not equal to expected (true)
 Arrays differ starting from [index: 7]:
 ... 5, 6,   7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, ...
 ... 5, 6, 125, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, ...
           ^^^^
 java.lang.RuntimeException: Result: (false) of 'arrayEqualsB' is not 
 equal to expected (true)
          at 
 compiler.intrinsics.string.TestStringIntrinsics.invokeAndCheckArrays(TestStringIntrinsics.java:273) 
           at ... stack trace continues - E.N.
```
 Sample output for comparing char arrays:
``` ----------System.err:(21/1579)*----------
 Result: (false) of 'arrayEqualsC' is not equal to expected (true)
 Arrays differ starting from [index: 7]:
 ... \\u0005, \\u0006, \\u0007, \\u0008, \\u0009, \\n, ...
 ... \\u0005, \\u0006,      }, \\u0008, \\u0009, \\n, ...
                     ^^^^^^^
 java.lang.RuntimeException: Result: (false) of 'arrayEqualsC' is not 
 equal to expected (true)
          at 
 compiler.intrinsics.string.TestStringIntrinsics.invokeAndCheckArrays(TestStringIntrinsics.java:280) 
           at
 ... and so on - E.N.
```

testing: open/test/hotspot/jtreg/compiler/intrinsics/string/TestStringIntrinsics.java on linux, windows, macosx.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8229186](https://bugs.openjdk.java.net/browse/JDK-8229186): Improve error messages for TestStringIntrinsics failures

### Reviewers
 * [Igor Ignatyev](https://openjdk.java.net/census#iignatyev) (@iignatev - **Reviewer**)
 * [Leonid Mesnik](https://openjdk.java.net/census#lmesnik) (@lmesnik - Committer) ⚠️ Review applies to 81f7a9d9b8847c8862e1edc7955a575641051339

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/112/head:pull/112`
`$ git checkout pull/112`

### CODE REVIEW

**Comment by lepestock** [comment]
Responding to the comments from pre-Skara thread:

> test/hotspot/jtreg/compiler/intrinsics/string/TestStringIntrinsics.java:
>    I'd prefer invokeAndCompareArrays and invokeAndCheck to be as close as possible: have both of them to accept either boolean or Object as 2nd arg; print/throw the same error message

the invokeAndCheck is very generic, it can be called with different objects and expect any kind of result, not only boolean. Therefore its output format radically differs from what an array-comparator should show. 

>  maybe I'm missing smth, but I don't understand why ArrayCodec supports only char and byte arrays; and hence I don't understand why you need ArrayCodec::of methods, as you can simply do new ArrayCoded(Arrays.stream(a).collect(Collectors.toList()) where a is an array of any type

for Object arrays, one can use that.
for integer primitives one needs Arrays.stream(a).boxed.collect(Collectors.toList()), please note 'boxed' - it is required and not generic.
for bytes or chars, there is none (no overload methos in the Arrays.stream(a));
To sum up, I can't see how with the given type system and utilities set I can make in a better, less wordy way. I've added int and long overloads, support for String and Object arrays to make it more complete.

> it seems that ArrayCodec should be an inner static class of ArrayDiff

I would argue that - I find it useful for printing arrays (and this usage has been utilised in the TestStringIntrinsics.java). In addition, I dont' like the practice of making such huge classes an inner classes as this reduces readability and modularity.

Other issues have been fixed. I added support for int, long, Object and String arrays.

**Comment by iignatev** [comment]
> Responding to the comments from pre-Skara thread:
> 
> > test/hotspot/jtreg/compiler/intrinsics/string/TestStringIntrinsics.java:
> > I'd prefer invokeAndCompareArrays and invokeAndCheck to be as close as possible: have both of them to accept either boolean or Object as 2nd arg; print/throw the same error message
> 
> the invokeAndCheck is very generic, it can be called with different objects and expect any kind of result, not only boolean. Therefore its output format radically differs from what an array-comparator should show.

I am not sure I understand what you mean... 
1. granted you can't change `invokeAndCheck`'s 2nd argument to bool as there are other values being passed, but you can change  `invokeAndCompareArrays` to accept an `Object` and compare expected and actual values by `Object::equals`.
2. even if you can't change output of these two methods to be the same (which I so far failed to see why), you still can change `invokeAndCheck`'s `message` var to include actual and expected values in the same way as `invokeAndCompareArrays` does.

> > maybe I'm missing smth, but I don't understand why ArrayCodec supports only char and byte arrays; and hence I don't understand why you need ArrayCodec::of methods, as you can simply do new ArrayCoded(Arrays.stream(a).collect(Collectors.toList()) where a is an array of any type
> 
> for Object arrays, one can use that.
> for integer primitives one needs Arrays.stream(a).boxed.collect(Collectors.toList()), please note 'boxed' - it is required and not generic.
> for bytes or chars, there is none (no overload methos in the Arrays.stream(a));
> To sum up, I can't see how with the given type system and utilities set I can make in a better, less wordy way. I've added int and long overloads, support for String and Object arrays to make it more complete.

you don't need `ArrayCodec::of(Object array)` anymore, do you?

> > it seems that ArrayCodec should be an inner static class of ArrayDiff
> 
> I would argue that - I find it useful for printing arrays (and this usage has been utilised in the TestStringIntrinsics.java). In addition, I dont' like the practice of making such huge classes an inner classes as this reduces readability and modularity.

oki.
> Other issues have been fixed. I added support for int, long, Object and String arrays.

File: test/lib/jdk/test/lib/format/ArrayDiff.java [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+
+    private final ArrayCodec<E> first;
+    private final ArrayCodec<E> second;
+
+    /**
+     * Default limits for the formatter
+     */
+    public static class Defaults {
+        final static int WIDTH = 80;
+        final static int CONTEXT_BEFORE = 2;
```
Comment (iignatev) [inline test/lib/jdk/test/lib/format/ArrayDiff.java:83]: either these constants should be `public`, or `Defaults` class should be `package-private`, otherwise, you get a public class w/ no public fields.

**Review by iignatev — CHANGES_REQUESTED** [review]

File: test/lib/jdk/test/lib/format/ArrayCodec.java [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+     * two strings have the same length;
+     *
+     * @param another Another codec to compare encoded element width with
+     */
+    public void alignBy(ArrayCodec<E> another) {
+        if (!element.equals("") && !element.equals("]")) {
+            int delta = another.element.length() - element.length();
+            if (delta > 0) {
+                element = Format.paddingForWidth(delta) + element;
+        }
```
[the remark this answers is not in the document]
Comment (lepestock) [reply test/lib/jdk/test/lib/format/ArrayCodec.java]: Fixed.

File: test/lib/jdk/test/lib/format/ArrayDiff.java (hunk 1 above)
Comment (lepestock) [reply test/lib/jdk/test/lib/format/ArrayDiff.java:83]: Old cpp sins... fixed.

File: test/lib-test/jdk/test/lib/format/ArrayDiffTest.java [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+    public void testFirstElementIsWrong() {
+        byte[] first = new byte[]  {122};
+        byte[] second = new byte[] {7, 8, 9, 10, 125, 12, 13};
+
+        ArrayDiff diff = ArrayDiff.of(first, second);
+        String expected = String.format(
+                "Arrays differ starting from [index: 0]:%n" +
+                        "[122]%n" +
+                        "[  7, 8, 9, 10, 125, 12, 13]%n" +
+                        " ^^^");
```
[the remark this answers is not in the document]
Comment (lepestock) [reply test/lib-test/jdk/test/lib/format/ArrayDiffTest.java:106]: ... in almost every test :). Fixed.

**Comment by lepestock** [comment]
> 1. granted you can't change `invokeAndCheck`'s 2nd argument to bool as there are other values being passed, but you can change  `invokeAndCompareArrays` to accept an `Object` and compare expected and actual values by `Object::equals`.

Well, I think that would make `invokeAndCompareArrays` look less specific and confuse a reader by creating a false impression that it could be called with an Object as an expected result. In reality, the method is only called with 'equals' of different signatures, and 'equals' does always return boolean. It could've been named 'invokeEqualsOnArraysAndCompareThem` :)

The other method, `invokeAndCheck` is different. It can call, for example, `String.concat("abc", "def")` and expect the String "abcdef" as result. It does really need to be generic.

> 2. even if you can't change output of these two methods to be the same (which I so far failed to see why), you still can change `invokeAndCheck`'s `message` var to include actual and expected values in the same way as `invokeAndCompareArrays` does.

`invokeAndCompareArrays` / `ArrayDiff` compares two huge arrays and present a nice short *slice* in the difference area. The short slice is possible because we do  a limited task - compare arrays.
 `invokeAndCheck` , on the other hand, can have as parameters the `utf16`, a string of 10K symbols. Do we really need a 10K string as output in our log?

> you don't need `ArrayCodec::of(Object array)` anymore, do you?

Unfortunately, it is used in the ArrayCodec.format (which is used in TestStringIntrinsics.java) - to make it possible to call it with everything and not swamp the code with overloads.

**Comment by iignatev** [comment]
> The other method, `invokeAndCheck` is different. It can call, for example, `String.concat("abc", "def")` and expect the String "abcdef" as result. It does really need to be generic.
although I still don't see them as really being that much different, I'm fine w/ keeping them as-is.


> `invokeAndCompareArrays` / `ArrayDiff` compares two huge arrays and present a nice short _slice_ in the difference area. The short slice is possible because we do a limited task - compare arrays.
> `invokeAndCheck` , on the other hand, can have as parameters the `utf16`, a string of 10K symbols. Do we really need a 10K string as output in our log?

I think we do, as people tend to first look at exception's messages and only then look thru other logs, so having relative information in the exception is always a good thing. should 10K symbols become a problem (this though would also mean that you can't print the compared values w/  `Format.asLiteral`), we can revisit this.

> 
> > you don't need `ArrayCodec::of(Object array)` anymore, do you?
> 
> Unfortunately, it is used in the ArrayCodec.format (which is used in TestStringIntrinsics.java) - to make it possible to call it with everything and not swamp the code with overloads.

hm, I somehow missed that usage, but you don't need to repeat to the same switch over a component type in `ArrayDiff::of`, do you?

File: test/lib/jdk/test/lib/format/ArrayCodec.java [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+        } else if (type == int.class) {
+            return ArrayCodec.of((int[])array);
+        } else if (type == long.class) {
+            return ArrayCodec.of((long[])array);
+        } else if (type == char.class) {
+            return ArrayCodec.of((char[])array);
+        } else if (type == String.class) {
+            return ArrayCodec.of((String[])array);
+        } else if (!type.isPrimitive() && !type.isArray()) {
+            return ArrayCodec.of((String[])array);
```
Comment (iignatev) [inline test/lib/jdk/test/lib/format/ArrayCodec.java:157]:
```suggestion
            return ArrayCodec.of((Object[])array);
```

copy-paste typo?

Comment (lepestock) [reply test/lib/jdk/test/lib/format/ArrayCodec.java]: A good one, thanks. Fixed.

**Comment by lepestock** [comment]
> I think we do, as people tend to first look at exception's messages and only then look thru other logs, so having relative information in the exception is always a good thing. should 10K symbols become a problem (this though would also mean that you can't print the compared values w/ `Format.asLiteral`), we can revisit this.

Ok, I got it. I was blind and stupid, sorry. Fixed, please check the e6fb6d04cad

> hm, I somehow missed that usage, but you don't need to repeat to the same switch over a component type in `ArrayDiff::of`, do you?

&lt;sigh&gt;... I can see no better solution here. The `ArrayDiff::of` checks that both component types are the same in addition, but I probably could have avoided duplicating the switch (we can ignore double guessing the type). The issue is that <E> in ArrayDiff is forwarded to first and second, I would need to declare them as generic (ArrayCodec<Object>). The type system seem to not allow me to fix that nicely.

**Comment by iignatev** [comment]
> > I think we do, as people tend to first look at exception's messages and only then look thru other logs, so having relative information in the exception is always a good thing. should 10K symbols become a problem (this though would also mean that you can't print the compared values w/ `Format.asLiteral`), we can revisit this.
> 
> Ok, I got it. I was blind and stupid, sorry. Fixed, please check the [e6fb6d0](https://github.com/openjdk/jdk/commit/e6fb6d04caddc4b410a594574f570cfcb9445e4a)

I'd use `StringBuilder` to construct the message.

> > hm, I somehow missed that usage, but you don't need to repeat to the same switch over a component type in `ArrayDiff::of`, do you?
>
> <sigh>... I can see no better solution here. The `ArrayDiff::of` checks that both component types are the same in addition, but I probably could have avoided duplicating the switch (we can ignore double guessing the type). The issue is that in ArrayDiff is forwarded to first and second, I would need to declare them as generic (ArrayCodec). The type system seem to not allow me to fix that nicely.

wouldn't the following patch do?


```diff --git a/test/lib/jdk/test/lib/format/ArrayDiff.java b/test/lib/jdk/test/lib/format/ArrayDiff.java
--- a/test/lib/jdk/test/lib/format/ArrayDiff.java
+++ b/test/lib/jdk/test/lib/format/ArrayDiff.java
@@ -123,41 +123,10 @@ public class ArrayDiff<E> implements Diff {
         if (!bothAreArrays || !componentTypesAreSame) {
             throw new IllegalArgumentException("Both arguments should be arrays of the same type");
         }
-
-        var type = first.getClass().getComponentType();
-        if (type == byte.class) {
-            return new ArrayDiff<>(
-                    ArrayCodec.of((byte[])first),
-                    ArrayCodec.of((byte[])second),
-                    width, contextBefore);
-        } else if (type == int.class) {
-            return new ArrayDiff<>(
-                    ArrayCodec.of((int[])first),
-                    ArrayCodec.of((int[])second),
-                    width, contextBefore);
-        } else if (type == long.class) {
-            return new ArrayDiff<>(
-                    ArrayCodec.of((long[])first),
-                    ArrayCodec.of((long[])second),
-                    width, contextBefore);
-        } else if (type == char.class) {
-            return new ArrayDiff<>(
-                    ArrayCodec.of((char[])first),
-                    ArrayCodec.of((char[])second),
-                    width, contextBefore);
-        } else if (type == String.class) {
-            return new ArrayDiff<>(
-                    ArrayCodec.of((String[])first),
-                    ArrayCodec.of((String[])second),
-                    width, contextBefore);
-        } else if (!type.isPrimitive() && !type.isArray()) {
-            return new ArrayDiff<>(
-                    ArrayCodec.of((Object[])first),
-                    ArrayCodec.of((Object[])second),
-                    width, contextBefore);
-        }
-
-        throw new IllegalArgumentException("Unsupported array component type: " + type);
+        return new ArrayDiff(
+                ArrayCodec.of(first),
+                ArrayCodec.of(second),
+                width, contextBefore);
     }
 
     /**
```

**Comment by lepestock** [comment]
> I'd use `StringBuilder` to construct the message.

Fixed, please check the f0b1df8.

> wouldn't the following patch do?

Wow, that worked. That type erasure... pushed in the same commit f0b1df8.

**Review by iignatev — APPROVED** [review]
LGTM

File: test/lib/jdk/test/lib/format/Format.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2015, 2020, Oracle and/or its affiliates. All rights reserved.
```
Comment (lmesnik) [inline test/lib/jdk/test/lib/format/Format.java:2]: Shouldn't be just 2020?

File: test/lib/jdk/test/lib/format/Diff.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+ * questions.
+ */
+
+/*
+ * Represents a possible difference between two objects.
+ */
+package jdk.test.lib.format;
+
+/**
+ * An abstraction representing formattabe difference between two or more objects
```
Comment (lmesnik) [inline test/lib/jdk/test/lib/format/Diff.java:30]: typo in 'formattabe'

File: test/lib/jdk/test/lib/format/ArrayDiff.java [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    private final int maxWidth;
+    private final int contextBefore;
+
+    private final ArrayCodec<E> first;
+    private final ArrayCodec<E> second;
+
+    /**
+     * Default limits for the formatter
+     */
+    public static class Defaults {
```
Comment (lmesnik) [inline test/lib/jdk/test/lib/format/ArrayDiff.java:81]: Does it make sense to move them into Diff? Seems like very generic formatter properties to me.

File: test/lib-test/jdk/test/lib/format/ArrayDiffTest.java [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.assertEquals;
+
+/*
+ * @test
+ * @summary Check ArrayDiff formatting
+ * @library /test/lib
+ * @run testng jdk.test.lib.format.ArrayDiffTest
+ */
+public class ArrayDiffTest {
```
Comment (lmesnik) [inline test/lib-test/jdk/test/lib/format/ArrayDiffTest.java:38]: Might be it makes sense to check null values. Also, reversed checks. For example there is first = [1, ..], second = []. But no check for  first = [], second = [1, ..] for some cases,

**Review by lmesnik — CHANGES_REQUESTED** [review]

File: test/lib/jdk/test/lib/format/Format.java (hunk 5 above)
Comment (lepestock) [reply test/lib/jdk/test/lib/format/Format.java]: Fixed in the d792e6d0ab9, thanks.

File: test/lib/jdk/test/lib/format/Diff.java (hunk 6 above)
Comment (lepestock) [reply test/lib/jdk/test/lib/format/Diff.java]: Also fixed in the d792e6d0ab9, thanks.

File: test/lib/jdk/test/lib/format/ArrayDiff.java (hunk 7 above)
Comment (lepestock) [reply test/lib/jdk/test/lib/format/ArrayDiff.java]: I can imagine different defaults for different diff formatters, but they probably would be very similar. Let's try your suggestion, the 81f7a9d9b88.

**Review by lmesnik — APPROVED** [review]

File: test/lib-test/jdk/test/lib/format/ArrayDiffTest.java (hunk 8 above)
Comment (lepestock) [reply test/lib-test/jdk/test/lib/format/ArrayDiffTest.java:38]: Fixed in the 39992fcba97. That ended in finding a minor issue, actually ([ArrayDiff.java:180](https://github.com/openjdk/jdk/pull/112/commits/39992fcba979a2f6bf7ea73aff92461b13220ee2#diff-374ead0149950fce10be8d9d37cca320L180)). Sometimes reversing can be valuable, thanks!

**Comment by shipilev** [comment]
@iignatev, @lmesnik: would you like to sponsor this?

File: test/lib/jdk/test/lib/format/ArrayDiff.java [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+     * @return an ArrayDiff instance for the two arrays and formatting parameters provided
+     */
+    public static ArrayDiff of(Object first, Object second, int width, int contextBefore) {
+        if (first == null) {
+            throw new IllegalArgumentException("First array argument for ArrayDiff is null");
+        }
+
+        if (second == null) {
+            throw new IllegalArgumentException("Second array argument for ArrayDiff is null");
+        }
```
Comment (iignatev) [inline test/lib/jdk/test/lib/format/ArrayDiff.java:117]: it's more common (and less surprising) to throw NPE in such cases, preferably by using `Objects::requireNonNull`

File: test/lib/jdk/test/lib/format/Diff.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+/**
+ * An abstraction representing formattable difference between two or more objects
+ */
+public interface Diff {
+
+    /**
+     * Default limits for formatters
+     */
+    public static class Defaults {
```
Comment (iignatev) [inline test/lib/jdk/test/lib/format/Diff.java:37]: I'd add a no-op private no-arg ctor to show that this class isn't expected to be instantiated.

File: test/lib/jdk/test/lib/format/ArrayCodec.java [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+                }
+            }
+
+            if (!first.equals(second)) {
+                return result;
+            }
+
+        }
+
+        return 0;
```
Comment (iignatev) [inline test/lib/jdk/test/lib/format/ArrayCodec.java:325]: shouldn't it be -1?

**Review by iignatev — CHANGES_REQUESTED** [review]

File: test/lib/jdk/test/lib/format/ArrayDiff.java (hunk 9 above)
Comment (lepestock) [reply test/lib/jdk/test/lib/format/ArrayDiff.java]: Haven't known about the `requireNonNull`, thanks. I was thinking about the NPE myself, just wanted to show up which array was null. Fixed in the ba6e5411b4de5cf8b9230b7cdb7a518900eadc4b.

File: test/lib/jdk/test/lib/format/Diff.java (hunk 10 above)
Comment (lepestock) [reply test/lib/jdk/test/lib/format/Diff.java:37]: Got it, fixed in the 78ed79f38262fec1df3ffb7ba407dc792666d1fd.

File: test/lib/jdk/test/lib/format/ArrayCodec.java (hunk 11 above)
Comment (lepestock) [reply test/lib/jdk/test/lib/format/ArrayCodec.java]: An interesting notice, actually. It should be either `-1` or `result` depending on whether arrays are equal, or one array is a prefix of another. Fixed in the 174148ab41e1c03d46dbde8862865dbaf3138843.

**Review by iignatev — APPROVED** [review]
